### PR TITLE
Fix loc behaviour when key of incorrect type is used

### LINKED ISSUE #6823: [BUG] Passing incorrectly typed object to `Series.__getitem__` causes uninformative Numba exception
Reproducer:
```python
import cudf

my_series = cudf.Series([1, 2, 3], index=[0, 1, 2])
print(my_series['0'])
```
Throws:
```python-traceback
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/rgsl888/Projects/cython/cudf/python/cudf/cudf/core/series.py", line 925, in __getitem__
    return self.loc[arg]
  File "/home/rgsl888/Projects/cython/cudf/python/cudf/cudf/core/indexing.py", line 118, in __getitem__
    arg = self._loc_to_iloc(arg)
  File "/home/rgsl888/Projects/cython/cudf/python/cudf/cudf/core/indexing.py", line 138, in _loc_to_iloc
    arg, closest=False
  File "/home/rgsl888/Projects/cython/cudf/python/cudf/cudf/core/column/numerical.py", line 455, in find_first_value
    self.data_array_view, value, mask=self.mask
  File "/home/rgsl888/Projects/cython/cudf/python/cudf/cudf/utils/cudautils.py", line 185, in find_first
    found_col = find_index_of_val(arr, val, mask=mask, compare=compare)
  File "/home/rgsl888/Projects/cython/cudf/python/cudf/cudf/utils/cudautils.py", line 165, in find_index_of_val
    arr, val, found, arr.size
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/cuda/compiler.py", line 372, in __call__
    kernel = self.kernel.specialize(*args)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/cuda/compiler.py", line 882, in specialize
    self._bind, targetoptions)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/cuda/compiler.py", line 808, in __init__
    self.compile(sigs[0])
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/cuda/compiler.py", line 932, in compile
    **self.targetoptions)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/cuda/compiler.py", line 57, in compile_kernel
    cres = compile_cuda(pyfunc, types.void, args, debug=debug, inline=inline)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/cuda/compiler.py", line 46, in compile_cuda
    locals={})
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler.py", line 627, in compile_extra
    return pipeline.compile_extra(func)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler.py", line 363, in compile_extra
    return self._compile_bytecode()
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler.py", line 425, in _compile_bytecode
    return self._compile_core()
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler.py", line 405, in _compile_core
    raise e
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler.py", line 396, in _compile_core
    pm.run(self.state)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler_machinery.py", line 341, in run
    raise patched_exception
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler_machinery.py", line 332, in run
    self._runPass(idx, pass_inst, state)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler_machinery.py", line 291, in _runPass
    mutated |= check(pss.run_pass, internal_state)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/compiler_machinery.py", line 264, in check
    mangled = func(compiler_state)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/typed_passes.py", line 98, in run_pass
    raise_errors=self._raise_errors)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/typed_passes.py", line 70, in type_inference_stage
    infer.propagate(raise_errors=raise_errors)
  File "/home/rgsl888/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/numba/core/typeinfer.py", line 1071, in propagate
    raise errors[0]
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Internal error at <numba.core.typeinfer.CallConstraint object at 0x7f226c0831d0>.
module, class, method, function, traceback, frame, or code object was expected, got CPUDispatcher
During: resolving callee type: Function(<numba.cuda.compiler.DeviceFunctionTemplate object at 0x7f226e0eac10>)
During: typing of call at /home/rgsl888/Projects/cython/cudf/python/cudf/cudf/utils/cudautils.py (98)
Enable logging at debug level for details.
File "cudf/utils/cudautils.py", line 98:
def gpu_mark_found_int(arr, val, out, not_found):
    <source elided>
    if i < arr.size:
        if check_equals_int(arr[i], val):
```
Expected:
```python
import pandas as pd

my_series = pd.Series([1, 2, 3], index=[0, 1, 2])
print(my_series['0'])
```
```python-traceback
Traceback (most recent call last):
  File "/home/keith/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/indexes/base.py", line 2889, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas/_libs/index.pyx", line 70, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 94, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index_class_helper.pxi", line 93, in pandas._libs.index.Int64Engine._check_type
KeyError: '0'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/keith/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/series.py", line 882, in __getitem__
    return self._get_value(key)
  File "/home/keith/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/series.py", line 991, in _get_value
    loc = self.index.get_loc(label)
  File "/home/keith/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/indexes/base.py", line 2891, in get_loc
    raise KeyError(key) from err
KeyError: '0'
```

### PR DESCRIPTION
**Pull request by shwina**
* Fixes #6823
* Raise a `KeyError` similar to Pandas rather than an `IndexError` when loc fails
* Improve tests to compare more directly with Pandas behaviour

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,6 +25,7 @@
 - PR #6942 Fix cudf::merge gtest for dictionary columns
 
 
+# cuDF 0.17.0 (Date TBD)
```
Comment (rgsl888prabhu) [inline CHANGELOG.md:28]: Merge issue?

File: python/cudf/cudf/core/column/numerical.py [hunk 2]
```diff
@@ -448,6 +449,8 @@ def find_first_value(self, value, closest=False):
         columns, returns the offset of the first larger value
         if closest=True.
         """
+        if not pd.api.types.is_number(value):
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/core/column/numerical.py:452]: May be we have to handle 0-d array as well.

```
>>> import pandas as pd
>>> import numpy as np
>>> pd.Series([1, 2, 3, 4, 5])[np.array(3)]
4

```

File: python/cudf/cudf/core/column/numerical.py [hunk 3]
```diff
@@ -474,6 +477,8 @@ def find_last_value(self, value, closest=False):
         columns, returns the offset of the last smaller value
         if closest=True.
         """
+        if not pd.api.types.is_number(value):
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/core/column/numerical.py:480]: similarly support for 0-d array.

File: python/cudf/cudf/core/column/numerical.py (hunk 2 above)
Comment (shwina) [reply python/cudf/cudf/core/column/numerical.py:452]: `np.array([3])` is a 1-dim array. `np.array([3])[0]` is a 0-dim array. It looks like both just work right now:

```python
In [4]: a = cudf.Series([1, 2, 3])

In [5]: a[np.array([0])]
Out[5]:
0    1
dtype: int64

In [6]: a[np.array([0])[0]]
Out[6]: 1
```

File: CHANGELOG.md (hunk 1 above)
Comment (shwina) [reply CHANGELOG.md]: Fixed!

File: python/cudf/cudf/core/column/numerical.py (hunk 2 above)
Comment (rgsl888prabhu) [reply python/cudf/cudf/core/column/numerical.py:452]: @shwina try with `np.array(0)`

Comment (shwina) [reply python/cudf/cudf/core/column/numerical.py:454]: Sorry! I misread. I added handling 0-dim arrays and a test!

**Review by rgsl888prabhu — APPROVED** [review]
LGTM

File: python/cudf/cudf/tests/test_indexing.py [hunk 4]
```diff
@@ -1300,3 +1319,27 @@ def test_iloc_with_lists(data, key):
     psr = pd.Series(data)
     gsr = cudf.Series(data)
     assert_eq(psr.iloc[key], gsr.iloc[key])
+
+
+@pytest.mark.parametrize("key", [5, -10, "0", "a"])
```
Comment (isVoid) [inline python/cudf/cudf/tests/test_indexing.py:1324]: I think for `bad_key` tests, there could be a test case for 0-d numpy array too? For example at this line, add `np.array(5)`, `np.array("a")`

File: python/cudf/cudf/tests/test_indexing.py [hunk 5]
```diff
@@ -1300,3 +1319,27 @@ def test_iloc_with_lists(data, key):
...
+def test_loc_bad_key_type(key):
+    psr = pd.Series([1, 2, 3])
+    gsr = cudf.from_pandas(psr)
+    assert_exceptions_equal(lambda: psr[key], lambda: gsr[key])
+    assert_exceptions_equal(lambda: psr.loc[key], lambda: gsr.loc[key])
+
+
+@pytest.mark.parametrize("key", ["b", 1.0])
```
Comment (isVoid) [inline python/cudf/cudf/tests/test_indexing.py:1332]: Same as above.

**Review by isVoid — COMMENTED** [review]
Test case qst, otherwise lgtm!

File: python/cudf/cudf/tests/test_indexing.py (hunk 4 above)
Comment (shwina) [reply python/cudf/cudf/tests/test_indexing.py:1324]: Done! Ditto below

**Review by isVoid — APPROVED** [review]
lgtm!

**Review by galipremsagar — APPROVED** [review]

**Pull request merged**